### PR TITLE
initial commit

### PR DESCRIPTION
**Pull request by ilyapozh**

### CODE REVIEW

File: api/src/filtersQueryPipe.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { ArgumentMetadata, Query, ValidationPipe } from '@nestjs/common';
+import { IsIn, IsInt, Min, IsString, MaxLength } from 'class-validator';
+
+export class NickQueryDto {
+    @IsString()
+    @MaxLength(20)
+    equal = '';
+}
+
+export class FiltersQueryPipe extends ValidationPipe {
```
Comment (tomfun) [inline api/src/filtersQueryPipe.ts:10]: Why we need `ValidationPipe`? Why we need to validate the query or the nick?

File: api/src/report.controller.ts [hunk 2]
```diff
@@ -18,9 +19,10 @@ export class ReportController {
   @Get('/api/report?')
   getReports(
     @PaginateQuery query: PaginationQueryDto,
-    @Query('nick') nick: string,
```
Comment (tomfun) [inline api/src/report.controller.ts:21]: Initial idea to create some decorator PaginationFilters or something and use it like that:
```ts
  @Get('/api/report')
  getReports(
    @PaginateQuery query: PaginationQueryDto,
    @PaginationFilters filters: PaginationFilter<Report>, // or `filters: ReportPaginationFilter`
  ): Promise<Paginable<ReportForList>> {
    return this.appService.getList(query,  filters ); // filters.nick and other fields in it
  }
```
But I don't remember when we wanted to do it, here in the current task?

File: my-vue3-app/src/components/ReportList.vue [hunk 3]
```diff
@@ -8,6 +8,26 @@
...
+        <label>{{filter.name}}</label>
+        <div v-if="filter.checked">
+          <select v-model="filter.filterType">
+            <option
+              v-for="(filType, index) in filterTypes"
+              v-bind:key="index">
+              {{filType.name}}
+            </option>
+          </select>
+          <input v-model="filter.inputValue" placeholder="edit me"/>
```
Comment (tomfun) [inline my-vue3-app/src/components/ReportList.vue:28]: Why always input?

File: my-vue3-app/src/components/ReportList.vue [hunk 4]
```diff
@@ -74,6 +116,16 @@ export default defineComponent({
     nick() {
       this.onNickInput();
     },
+    filters: {
+      handler(newValue, oldValue) {
```
Comment (tomfun) [inline my-vue3-app/src/components/ReportList.vue:120]: I will study the subject but not today :(

File: api/src/filtersQueryPipe.ts (hunk 1 above)
Comment (tomfun) [reply api/src/filtersQueryPipe.ts:10]: https://docs.nestjs.com/pipes#built-in-pipes

Comment (tomfun) [reply api/src/filtersQueryPipe.ts:10]: image

File: my-vue3-app/src/components/ReportList.vue (hunk 3 above)
Comment (ilyapozh) [reply my-vue3-app/src/components/ReportList.vue:28]: If not input then what ?

File: api/src/filtersQueryPipe.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    @Type(() => NickQueryDto)
+    nick: NickQueryDto
+    @ValidateNested()
+    @Type(() => TitleQueryDto)
+    title: TitleQueryDto
+}
+
+export class FiltersQueryPipe implements PipeTransform {
+    transform(value: any, metadata: ArgumentMetadata) {
+        let filters = {};
```
Comment (tomfun) [inline api/src/filtersQueryPipe.ts:29]: much better to set type here
```ts
let filters = {} as TitleQueryDto; // or other type. not sure
```

File: api/src/filtersQueryPipe.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        if (value.nick) {
+            Object.assign(filters, value.nick.equal ? {
+                nick: {
+                    equal: value.nick.equal
+                }
+            } : '')
```
Comment (tomfun) [inline api/src/filtersQueryPipe.ts:35]:
```ts
Object.assign(filters, '')
// ____________________^ very strange
```

File: api/src/filtersQueryPipe.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    equal = '';
+}
+
+export class TitleQueryDto {
+    @IsString()
+    @MaxLength(20)
+    equal = '';
+}
+
+export class PaginationFilter<ReportEntity> {
```
Comment (tomfun) [inline api/src/filtersQueryPipe.ts:18]: 'ReportEntity' is defined but never used

File: my-vue3-app/src/components/ReportList.vue (hunk 3 above)
Comment (tomfun) [reply my-vue3-app/src/components/ReportList.vue:26]: select!?
but yes, for now it is enough

File: my-vue3-app/src/store/reports/state.ts [hunk 8]
```diff
@@ -46,7 +53,15 @@ export default {
     ids: [],
     viewIds: [],
     filters: {
-      nick: '',
+      // nick: '',
+      nick: {
+        value: '',
+        type: '',
```
Comment (tomfun) [inline my-vue3-app/src/store/reports/state.ts:59]: we could use special type `'none'` or `null` for no filtering

File: package.json [hunk 9]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "run-all",
+  "version": "1.0.0",
+  "description": "## Project setup",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "start": "docker-compose up & (cd api && npm start) & (cd my-vue3-app && npm run serve)"
```
Comment (tomfun) [inline package.json:7]: Do you really want add this file because of this string?
Better to move this only line to readme

**Review by tomfun — CHANGES_REQUESTED** [review]

File: api/package-lock.json [hunk 10]
```diff
@@ -1,8 +1,10806 @@
...
+        "prettier": "^2.1.2",
+        "supertest": "^6.0.0",
+        "ts-jest": "^26.4.3",
+        "ts-loader": "^8.0.8",
+        "ts-node": "^9.0.0",
+        "tsconfig-paths": "^3.9.0",
+        "typescript": "^4.0.5"
+      }
+    },
+    "node_modules/@angular-devkit/core": {
```
Comment (tomfun) [inline api/package-lock.json:47]: why at all this file was updated? it seems it is updated by mistake

File: my-vue3-app/src/store/reports/state.ts (hunk 8 above)
Comment (tomfun) [reply my-vue3-app/src/store/reports/state.ts:59]: I used `value = undefined` for this case ...

File: my-vue3-app/src/components/ReportList.vue (hunk 4 above)
Comment (tomfun) [reply my-vue3-app/src/components/ReportList.vue]: seems we fixed it. but the problem with pagination left...

File: api/src/filtersQueryPipe.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    nick: NickQueryDto
+    @ValidateNested()
+    @Type(() => TitleQueryDto)
+    title: TitleQueryDto
+}
+
+export class FiltersQueryPipe implements PipeTransform {
+    transform(value: any, metadata: ArgumentMetadata) {
+        let filters = {};
+        if (value.nick && value.nick.equal) {
```
[the remark this answers is not in the document]
Comment (tomfun) [reply api/src/filtersQueryPipe.ts]: seems it is working well

File: api/src/filtersQueryPipe.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+          type: reportConfig[fieldName].type,
+          filters: {},
+        };
+        for (const operator of reportConfig[fieldName].filters) {
+          if (operator in v && typeof v[operator] === 'string') {
+            filters[fieldName].filters[operator] = v[operator] as string;
+          }
+        }
+      });
+    console.log('f2', filters);
```
Comment (tomfun) [inline api/src/filtersQueryPipe.ts:68]: rm

File: my-vue3-app/src/store/reports/actions.ts [hunk 13]
```diff
@@ -9,16 +9,31 @@ import {
...
   desired: Omit<Pagination, 'viewIds'|'ids'>,
 ): Array<string> => {
   const ids = Object.keys(reports);
+  const filterFields = ['nick' as 'nick', 'title' as 'title']
+    .filter((field) => desired.filters[field].value !== undefined);
   const filtered = desired.filters && desired.filters.nick
     ? ids
       .map((id) => reports[id])
-      .filter((report) => report.nick === desired.filters.nick)
+      .filter((report) => filterFields.reduce((ok, field) => ok
```
Comment (tomfun) [inline my-vue3-app/src/store/reports/actions.ts:34]: reduce -> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every

File: my-vue3-app/src/store/reports/actions.ts [hunk 14]
```diff
@@ -9,16 +9,31 @@ import {
...
+  }
+};
+
 const applyFilterSortToReports = (
   pagination: Pick<Pagination, 'order' | 'filters'>,
   reports: Reports,
   desired: Omit<Pagination, 'viewIds'|'ids'>,
 ): Array<string> => {
   const ids = Object.keys(reports);
+  const filterFields = ['nick' as 'nick', 'title' as 'title']
```
Comment (tomfun) [inline my-vue3-app/src/store/reports/actions.ts:29]:
```ts
'nick' as 'nick'
```
->
```ts
'nick' as const
```

File: package.json (hunk 9 above)
Comment (tomfun) [reply package.json]: We still may create `cheat sheet . md` with instructions

File: api/package-lock.json (hunk 10 above)
Comment (tomfun) [reply api/package-lock.json:51]: I fixed it

**Review by tomfun — CHANGES_REQUESTED** [review]
small fixes

**Review by tomfun — APPROVED** [review]

**Comment by ilyapozh** [comment]
There is an error on report submit with title or nick fields less then 4 characters, I think we should handle it on the front as well.

**Comment by ilyapozh** [comment]
valid
Maybe we should validate such nicks and titles somehow

**Comment by ilyapozh** [comment]
There are empty requests provided when selecting nick/title or equal/start

**Comment by ilyapozh** [comment]
Shouldn't we add the body of the report not hard-coded?

**Pull request merged**